compile problem IRremote

First of all, thank you for your program.

With IRremoteESP8266 in Version 2.3.2 I get the following errors:

`Arilux_AL-LC0X:80: error: 'IRrecv' does not name a type
 IRrecv            irRecv(ARILUX_IR_PIN);
 ^
`

And after downgrade to V1.2.0 I get some errors like this:

`/Volumes/myData/Documents/Arduino/libraries/IRremoteESP8266-1.2.0/IRKelvinator.cpp: In member function 'void IRKelvinatorAC::setTemp(uint8_t)':
/Volumes/myData/Documents/Arduino/libraries/IRremoteESP8266-1.2.0/IRKelvinator.cpp:96:41: error: no matching function for call to 'max(unsigned int, uint8_t&)'
     temp = max(KELVINATOR_MIN_TEMP, temp);
                                         ^
`

I hope, you can help me
thx

--- a/Arilux.cpp
+++ b/Arilux.cpp
@@ -71,7 +71,7 @@ uint8_t Arilux::increaseBrightness(void) {
   if (!m_state)
     return false;
   if ((m_brightness + ARILUX_BRIGHTNESS_STEP) >= ARILUX_PWM_RANGE)
-    return setBrightness(ARILUX_PWM_RANGE);  
+    return setBrightness(ARILUX_PWM_RANGE);
   if (m_brightness < (ARILUX_PWM_RANGE - ARILUX_BRIGHTNESS_STEP))
     return setBrightness(m_brightness + ARILUX_BRIGHTNESS_STEP);
   return false;
@@ -91,6 +91,9 @@ uint8_t Arilux::setBrightness(uint8_t p_brightness) {
     return false;
 
   m_brightness = p_brightness;
+
+  if (!m_state)
+    return false;
 
   analogWrite(m_redPin, map(m_color.red, 0, ARILUX_PWM_RANGE, 0, m_brightness));
   analogWrite(m_greenPin, map(m_color.green, 0, ARILUX_PWM_RANGE, 0, m_brightness));

--- a/Arilux_AL-LC0X.ino
+++ b/Arilux_AL-LC0X.ino
@@ -528,17 +528,18 @@ volatile unsigned long lastmqttreconnect = 0;
 void connectMQTT(void) {
   if (!mqttClient.connected()) {
     if (lastmqttreconnect + 1000 < millis()) {
-      if (mqttClient.connect(MQTT_CLIENT_ID, MQTT_USER, MQTT_PASS, ARILUX_MQTT_STATUS_TOPIC, 0, 1, "dead")) {
+      if (mqttClient.connect(MQTT_CLIENT_ID, MQTT_USER, MQTT_PASS, ARILUX_MQTT_STATUS_TOPIC, 0, 1, "offline")) {
         DEBUG_PRINTLN(F("INFO: The client is successfully connected to the MQTT broker"));
-        publishToMQTT(ARILUX_MQTT_STATUS_TOPIC, "alive");
+        publishToMQTT(ARILUX_MQTT_STATUS_TOPIC, "online");
         #ifdef HOME_ASSISTANT_MQTT_DISCOVERY
           DynamicJsonDocument root(JSON_OBJECT_SIZE(14));
           root["name"] = friendlyName;
+          root["availability_topic"] = ARILUX_MQTT_STATUS_TOPIC;
           #ifdef HOME_ASSISTANT_MQTT_ATTRIBUTES
             root["attributes_topic"] = HOME_ASSISTANT_MQTT_ATTRIBUTES_TOPIC;
           #endif
           #ifdef JSON
-            root["platform"] = "mqtt_json";
+            root["schema"] = "json";
             root["state_topic"] = ARILUX_MQTT_JSON_STATE_TOPIC;
             root["command_topic"] = ARILUX_MQTT_JSON_COMMAND_TOPIC;
             root["brightness"] = true;

--- a/Arilux_AL-LC0X.ino
+++ b/Arilux_AL-LC0X.ino
@@ -881,13 +881,16 @@ void handleCMD(void) {
         break;
       case ARILUX_CMD_BRIGHTNESS_CHANGED:
         publishBrightnessChange();
+        publishStateChange();
         break;
       case ARILUX_CMD_COLOR_CHANGED:
         publishColorChange();
+        publishStateChange();
         break;
       #if defined(RGBW) || defined (RGBWW)
         case ARILUX_CMD_WHITE_CHANGED:
           publishWhiteChange();
+          publishStateChange();
           break;
       #endif
       default:

--- a/Arilux_AL-LC0X.ino
+++ b/Arilux_AL-LC0X.ino
@@ -71,7 +71,7 @@ char   ARILUX_MQTT_STATUS_TOPIC[44];
 // MQTT buffer
 char msgBuffer[32];
 
-char friendlyName[32];
+char friendlyName[64];
 
 volatile uint8_t cmd = ARILUX_CMD_NOT_DEFINED;
 
@@ -191,6 +191,10 @@ void publishToMQTT(const char* topic, const char* payload) {
     #endif
     root["RSSI"] = WiFi.RSSI();
     root["SSID"] = WiFi.SSID();
+    root["Telnet Logging Enabled"] = false;
+    #if defined(DEBUG_TELNET)
+      root["Telnet Logging Enabled"] = true;
+    #endif
     char outgoingJsonBuffer[512];
     serializeJson(root, outgoingJsonBuffer);
     publishToMQTT(HOME_ASSISTANT_MQTT_ATTRIBUTES_TOPIC, outgoingJsonBuffer);
@@ -903,20 +907,20 @@ void handleCMD(void) {
 void setupWiFi() {
   delay(10);
 
-  Serial.print(F("INFO: Connecting to: "));
-  Serial.println(WIFI_SSID);
+  DEBUG_PRINT(F("INFO: Connecting to: "));
+  DEBUG_PRINTLN(WIFI_SSID);
 
   WiFi.mode(WIFI_STA);
   WiFi.begin(WIFI_SSID, WIFI_PASSWORD);
   while (WiFi.status() != WL_CONNECTED) {
     delay(500);
-    Serial.print(".");
+    DEBUG_PRINT(".");
   }
   randomSeed(micros());
-  Serial.println();
-  Serial.println(F("INFO: WiFi connected"));
-  Serial.print(F("INFO: IP address: "));
-  Serial.println(WiFi.localIP());
+  DEBUG_PRINTLN();
+  DEBUG_PRINTLN(F("INFO: WiFi connected"));
+  DEBUG_PRINT(F("INFO: IP address: "));
+  DEBUG_PRINTLN(WiFi.localIP());
 }
 
 ///////////////////////////////////////////////////////////////////////////
@@ -934,9 +938,9 @@ void setup() {
 
   sprintf(chipid, "%08X", ESP.getChipId());
   sprintf(MQTT_CLIENT_ID, HOST, chipid);
-  sprintf(friendlyName, "Arilux %s %s LED Controller %s", DEVICE_MODEL, arilux.getColorString(), chipid);
-  Serial.print("Hostname:");
-  Serial.println(MQTT_CLIENT_ID);
+  sprintf(friendlyName, "Arilux %s %s LED Controller (%s)", DEVICE_MODEL, arilux.getColorString(), chipid);
+  DEBUG_PRINT("Hostname:");
+  DEBUG_PRINTLN(MQTT_CLIENT_ID);
   WiFi.hostname(MQTT_CLIENT_ID);
 
   // Setup Wi-Fi

--- a/Arilux_AL-LC0X.ino
+++ b/Arilux_AL-LC0X.ino
@@ -72,11 +72,6 @@ char   ARILUX_MQTT_STATUS_TOPIC[44];
 char msgBuffer[32];
 
 char friendlyName[32];
-char discoveryBuf[512];
-char attributesBuf[512];
-#ifdef HOME_ASSISTANT_MQTT_DISCOVERY
-  StaticJsonDocument<512> HOME_ASSISTANT_MQTT_DISCOVERY_CONFIG;
-#endif
 
 volatile uint8_t cmd = ARILUX_CMD_NOT_DEFINED;
 
@@ -180,24 +175,61 @@ void publishToMQTT(const char* topic, const char* payload) {
 
 #ifdef HOME_ASSISTANT_MQTT_ATTRIBUTES
   void publishAttributes(void) {
-    DynamicJsonDocument root(9);
+    StaticJsonDocument<512> root;
+    root["BSSID"] = WiFi.BSSIDstr();
+    root["Chip ID"] = chipid;
     root["Hostname"] = MQTT_CLIENT_ID;
-    root["Model"] = DEVICE_MODEL;
-    root["IP Address"] = WiFi.localIP();
-    root["MAC Address"] = WiFi.macAddress();
-    root["RSSI"] = WiFi.RSSI();
-    root["SSID"] = WiFi.SSID();
-    root["BSSID"] = WiFi.BSSID();
+    root["IP Address"] = WiFi.localIP().toString();
     root["LED Strip Type"] = arilux.getColorString();
+    root["MAC Address"] = WiFi.macAddress();
+    root["Model"] = DEVICE_MODEL;
     root["Remote Type"] = "None";
     #if defined(IR_REMOTE)
       root["Remote Type"] = "Infrared (IR)";
     #elif defined(RF_REMOTE)
       root["Remote Type"] = "Radio Frequency (RF)";
     #endif
-    char outgoingJsonBuffer[12];
+    root["RSSI"] = WiFi.RSSI();
+    root["SSID"] = WiFi.SSID();
+    char outgoingJsonBuffer[512];
     serializeJson(root, outgoingJsonBuffer);
     publishToMQTT(HOME_ASSISTANT_MQTT_ATTRIBUTES_TOPIC, outgoingJsonBuffer);
+  }
+#endif
+
+
+#ifdef HOME_ASSISTANT_MQTT_DISCOVERY
+  void publishDiscovery(void) {
+    StaticJsonDocument<512> root;
+    root["name"] = friendlyName;
+    root["availability_topic"] = ARILUX_MQTT_STATUS_TOPIC;
+
+    #ifdef HOME_ASSISTANT_MQTT_ATTRIBUTES
+      root["json_attributes_topic"] = HOME_ASSISTANT_MQTT_ATTRIBUTES_TOPIC;
+    #endif
+
+    #ifdef JSON
+      root["brightness"] = true;
+      root["command_topic"] = ARILUX_MQTT_JSON_COMMAND_TOPIC;
+      root["rgb"] = true;
+      root["schema"] = "json";
+      root["state_topic"] = ARILUX_MQTT_JSON_STATE_TOPIC;
+      #if defined(RGBW) || defined (RGBWW)
+      root["white_value"] = true;
+      #endif
+    #else
+      root["brightness_command_topic"] = ARILUX_MQTT_BRIGHTNESS_COMMAND_TOPIC;
+      root["brightness_state_topic"] = ARILUX_MQTT_BRIGHTNESS_STATE_TOPIC;
+      root["command_topic"] = ARILUX_MQTT_STATE_COMMAND_TOPIC;
+      root["payload_off"] = MQTT_STATE_OFF_PAYLOAD;
+      root["payload_on"] = MQTT_STATE_ON_PAYLOAD;
+      root["rgb_command_topic"] = ARILUX_MQTT_COLOR_COMMAND_TOPIC;
+      root["rgb_state_topic"] = ARILUX_MQTT_COLOR_STATE_TOPIC;
+      root["state_topic"] = ARILUX_MQTT_STATE_STATE_TOPIC;
+    #endif
+    char outgoingJsonBuffer[1024];
+    serializeJson(root, outgoingJsonBuffer);
+    publishToMQTT(HOME_ASSISTANT_MQTT_DISCOVERY_TOPIC, outgoingJsonBuffer);
   }
 #endif
 
@@ -399,7 +431,7 @@ void callback(char* p_topic, byte* p_payload, unsigned int p_length) {
   // Handle the MQTT topic of the received message
   #ifdef JSON
     if (String(ARILUX_MQTT_JSON_COMMAND_TOPIC).equals(p_topic)) {
-      DynamicJsonDocument doc(1024);
+      StaticJsonDocument<512> doc;
       auto error = deserializeJson(doc, payload);
       if (error) {
         DEBUG_PRINT(F("parseObject() failed with code "));
@@ -531,34 +563,11 @@ void connectMQTT(void) {
       if (mqttClient.connect(MQTT_CLIENT_ID, MQTT_USER, MQTT_PASS, ARILUX_MQTT_STATUS_TOPIC, 0, 1, "offline")) {
         DEBUG_PRINTLN(F("INFO: The client is successfully connected to the MQTT broker"));
         publishToMQTT(ARILUX_MQTT_STATUS_TOPIC, "online");
+        #ifdef HOME_ASSISTANT_MQTT_ATTRIBUTES
+          publishAttributes();
+        #endif
         #ifdef HOME_ASSISTANT_MQTT_DISCOVERY
-          DynamicJsonDocument root(JSON_OBJECT_SIZE(14));
-          root["name"] = friendlyName;
-          root["availability_topic"] = ARILUX_MQTT_STATUS_TOPIC;
-          #ifdef HOME_ASSISTANT_MQTT_ATTRIBUTES
-            root["attributes_topic"] = HOME_ASSISTANT_MQTT_ATTRIBUTES_TOPIC;
-          #endif
-          #ifdef JSON
-            root["schema"] = "json";
-            root["state_topic"] = ARILUX_MQTT_JSON_STATE_TOPIC;
-            root["command_topic"] = ARILUX_MQTT_JSON_COMMAND_TOPIC;
-            root["brightness"] = true;
-            root["rgb"] = true;
-            #if defined(RGBW) || defined (RGBWW)
-            root["white_value"] = true;
-            #endif
-          #else
-            root["state_topic"] = ARILUX_MQTT_STATE_STATE_TOPIC;
-            root["command_topic"] = ARILUX_MQTT_STATE_COMMAND_TOPIC;
-            root["brightness_state_topic"] = ARILUX_MQTT_BRIGHTNESS_STATE_TOPIC;
-            root["brightness_command_topic"] = ARILUX_MQTT_BRIGHTNESS_COMMAND_TOPIC;
-            root["rgb_state_topic"] = ARILUX_MQTT_COLOR_STATE_TOPIC;
-            root["rgb_command_topic"] = ARILUX_MQTT_COLOR_COMMAND_TOPIC;
-            root["payload_on"] = MQTT_STATE_ON_PAYLOAD;
-            root["payload_off"] = MQTT_STATE_OFF_PAYLOAD;
-          #endif
-          serializeJson(root, discoveryBuf);
-          publishToMQTT(HOME_ASSISTANT_MQTT_DISCOVERY_TOPIC, discoveryBuf);
+          publishDiscovery();
         #endif
         flashSuccess(true);
       } else {
@@ -569,6 +578,8 @@ void connectMQTT(void) {
         DEBUG_PRINTLN(MQTT_PASS);
         DEBUG_PRINT(F("Broker: "));
         DEBUG_PRINTLN(MQTT_SERVER);
+        DEBUG_PRINT(F("Port: "));
+        DEBUG_PRINTLN(MQTT_PORT);
         flashSuccess(false);
       }
 
@@ -843,7 +854,7 @@ void handleRFRemote(void) {
 void handleCMD(void) {
   #ifdef JSON
   if (cmd != ARILUX_CMD_NOT_DEFINED) {
-      DynamicJsonDocument root(JSON_OBJECT_SIZE(7));
+      StaticJsonDocument<512> root;
       String stringState = arilux.getState() ? "ON" : "OFF";
       root["state"] = stringState;
       root["brightness"] = arilux.getBrightness();
@@ -853,7 +864,7 @@ void handleCMD(void) {
       color["r"] = arilux.getRedValue();
       color["g"] = arilux.getGreenValue();
       color["b"] = arilux.getBlueValue();
-      char outgoingJsonBuffer[12];
+      char outgoingJsonBuffer[512];
       serializeJson(root, outgoingJsonBuffer);
       publishToMQTT(ARILUX_MQTT_JSON_STATE_TOPIC, outgoingJsonBuffer);
   };
@@ -959,7 +970,7 @@ void setup() {
 #endif
 
 #ifdef HOME_ASSISTANT_MQTT_DISCOVERY
-  sprintf(HOME_ASSISTANT_MQTT_DISCOVERY_TOPIC,"%s/light/ARILUX_%s_%s_%s/config",HOME_ASSISTANT_MQTT_DISCOVERY_PREFIX,DEVICE_MODEL,arilux.getColorString(),chipid);
+  sprintf(HOME_ASSISTANT_MQTT_DISCOVERY_TOPIC, "%s/light/ARILUX_%s_%s_%s/config", HOME_ASSISTANT_MQTT_DISCOVERY_PREFIX, DEVICE_MODEL, arilux.getColorString(), chipid);
 #endif
 
 #ifdef JSON

--- a/Arilux_AL-LC0X.ino
+++ b/Arilux_AL-LC0X.ino
@@ -14,7 +14,7 @@
 #include <RCSwitch.h>         // https://github.com/sui77/rc-switch
 #endif
 #include <ArduinoOTA.h>
-#if defined(HOME_ASSISTANT_MQTT_DISCOVERY) || defined (JSON)
+#if defined(HOME_ASSISTANT_MQTT_DISCOVERY) || defined(HOME_ASSISTANT_MQTT_ATTRIBUTES) || defined (JSON)
   #include <ArduinoJson.h>
 #endif
 #include "Arilux.h"
@@ -44,6 +44,9 @@ char   MQTT_TOPIC_PREFIX[32];
 
 // MQTT topics
 char   ARILUX_MQTT_STATUS_TOPIC[44];
+#ifdef HOME_ASSISTANT_MQTT_ATTRIBUTES
+  char   HOME_ASSISTANT_MQTT_ATTRIBUTES_TOPIC[44];
+#endif
 #ifdef HOME_ASSISTANT_MQTT_DISCOVERY
   char   HOME_ASSISTANT_MQTT_DISCOVERY_TOPIC[56];
 #endif
@@ -68,7 +71,11 @@ char msgBuffer[32];
 char outgoingJsonBuffer[120];
 
 char friendlyName[32];
-char configBuf[512];
+char discoveryBuf[512];
+char attributesBuf[512];
+#ifdef HOME_ASSISTANT_MQTT_ATTRIBUTES
+  StaticJsonBuffer<512> HOME_ASSISTANT_MQTT_ATTRIBUTES_CONFIG;
+#endif
 #ifdef HOME_ASSISTANT_MQTT_DISCOVERY
   StaticJsonBuffer<512> HOME_ASSISTANT_MQTT_DISCOVERY_CONFIG;
 #endif
@@ -172,6 +179,28 @@ void publishToMQTT(const char* topic, const char* payload) {
     DEBUG_PRINTLN(F("ERROR: MQTT message publish failed, either connection lost, or message too large"));
   }
 }
+
+#ifdef HOME_ASSISTANT_MQTT_ATTRIBUTES
+  void publishAttributes(void) {
+    JsonObject& root = HOME_ASSISTANT_MQTT_ATTRIBUTES_CONFIG.createObject();
+    root["Hostname"] = MQTT_CLIENT_ID;
+    root["Model"] = DEVICE_MODEL;
+    root["IP Address"] = WiFi.localIP();
+    root["MAC Address"] = WiFi.macAddress();
+    root["RSSI"] = WiFi.RSSI();
+    root["SSID"] = WiFi.SSID();
+    root["BSSID"] = WiFi.BSSID();
+    root["LED Strip Type"] = arilux.getColorString();
+    root["Remote Type"] = "None";
+    #if defined(IR_REMOTE)
+      root["Remote Type"] = "Infrared (IR)";
+    #elif defined(RF_REMOTE)
+      root["Remote Type"] = "Radio Frequency (RF)";
+    #endif
+    root.printTo(attributesBuf, sizeof(attributesBuf));
+    publishToMQTT(HOME_ASSISTANT_MQTT_ATTRIBUTES_TOPIC, attributesBuf);
+  }
+#endif
 
 #ifndef JSON
   void publishStateChange(void) {
@@ -324,7 +353,7 @@ void handleEffects(void) {
       stepG = calculateStep(grnVal, realGreen);
       stepB = calculateStep(bluVal, realBlue);
       arilux.setFadeToColor(realRed, realGreen, realBlue);
-      
+
       inFade = true;
     }
   }
@@ -504,6 +533,9 @@ void connectMQTT(void) {
         #ifdef HOME_ASSISTANT_MQTT_DISCOVERY
           JsonObject& root = HOME_ASSISTANT_MQTT_DISCOVERY_CONFIG.createObject();
           root["name"] = friendlyName;
+          #ifdef HOME_ASSISTANT_MQTT_ATTRIBUTES
+            root["attributes_topic"] = HOME_ASSISTANT_MQTT_ATTRIBUTES_TOPIC;
+          #endif
           #ifdef JSON
             root["platform"] = "mqtt_json";
             root["state_topic"] = ARILUX_MQTT_JSON_STATE_TOPIC;
@@ -523,8 +555,8 @@ void connectMQTT(void) {
             root["payload_on"] = MQTT_STATE_ON_PAYLOAD;
             root["payload_off"] = MQTT_STATE_OFF_PAYLOAD;
           #endif
-          root.printTo(configBuf, sizeof(configBuf));
-          publishToMQTT(HOME_ASSISTANT_MQTT_DISCOVERY_TOPIC, configBuf);
+          root.printTo(discoveryBuf, sizeof(discoveryBuf));
+          publishToMQTT(HOME_ASSISTANT_MQTT_DISCOVERY_TOPIC, discoveryBuf);
         #endif
         flashSuccess(true);
       } else {
@@ -919,6 +951,10 @@ void setup() {
   sprintf(MQTT_TOPIC_PREFIX, MQTT_TOPIC_PREFIX_TEMPLATE, arilux.getColorString(), chipid);
 
   sprintf(ARILUX_MQTT_STATUS_TOPIC, MQTT_STATUS_TOPIC_TEMPLATE, MQTT_TOPIC_PREFIX);
+
+#ifdef HOME_ASSISTANT_MQTT_ATTRIBUTES
+  sprintf(HOME_ASSISTANT_MQTT_ATTRIBUTES_TOPIC, HOME_ASSISTANT_MQTT_ATTRIBUTES_TOPIC_TEMPLATE, MQTT_TOPIC_PREFIX);
+#endif
 
 #ifdef HOME_ASSISTANT_MQTT_DISCOVERY
   sprintf(HOME_ASSISTANT_MQTT_DISCOVERY_TOPIC,"%s/light/ARILUX_%s_%s_%s/config",HOME_ASSISTANT_MQTT_DISCOVERY_PREFIX,DEVICE_MODEL,arilux.getColorString(),chipid);

--- a/Arilux_AL-LC0X.ino
+++ b/Arilux_AL-LC0X.ino
@@ -6,16 +6,18 @@
 
 #include "config.h"
 #include <ESP8266WiFi.h>        // https://github.com/esp8266/Arduino
-#include <PubSubClient.h>       // https://github.com/knolleary/pubsubclient/releases/tag/v2.6
+#include <PubSubClient.h>       // https://github.com/knolleary/pubsubclient - v2.7
 #ifdef IR_REMOTE
-#include <IRremoteESP8266.h>  // https://github.com/markszabo/IRremoteESP8266
+#include <IRremoteESP8266.h>    // https://github.com/markszabo/IRremoteESP8266 - v2.5.4
+#include <IRrecv.h>
+#include <IRutils.h>
 #endif
 #ifdef RF_REMOTE
-#include <RCSwitch.h>         // https://github.com/sui77/rc-switch
+#include <RCSwitch.h>           // https://github.com/sui77/rc-switch
 #endif
 #include <ArduinoOTA.h>
 #if defined(HOME_ASSISTANT_MQTT_DISCOVERY) || defined(HOME_ASSISTANT_MQTT_ATTRIBUTES) || defined (JSON)
-  #include <ArduinoJson.h>
+  #include <ArduinoJson.h>      // https://github.com/bblanchon/ArduinoJson - Version 6.9.1
 #endif
 #include "Arilux.h"
 
@@ -68,16 +70,12 @@ char   ARILUX_MQTT_STATUS_TOPIC[44];
 
 // MQTT buffer
 char msgBuffer[32];
-char outgoingJsonBuffer[120];
 
 char friendlyName[32];
 char discoveryBuf[512];
 char attributesBuf[512];
-#ifdef HOME_ASSISTANT_MQTT_ATTRIBUTES
-  StaticJsonBuffer<512> HOME_ASSISTANT_MQTT_ATTRIBUTES_CONFIG;
-#endif
 #ifdef HOME_ASSISTANT_MQTT_DISCOVERY
-  StaticJsonBuffer<512> HOME_ASSISTANT_MQTT_DISCOVERY_CONFIG;
+  StaticJsonDocument<512> HOME_ASSISTANT_MQTT_DISCOVERY_CONFIG;
 #endif
 
 volatile uint8_t cmd = ARILUX_CMD_NOT_DEFINED;
@@ -182,7 +180,7 @@ void publishToMQTT(const char* topic, const char* payload) {
 
 #ifdef HOME_ASSISTANT_MQTT_ATTRIBUTES
   void publishAttributes(void) {
-    JsonObject& root = HOME_ASSISTANT_MQTT_ATTRIBUTES_CONFIG.createObject();
+    DynamicJsonDocument root(9);
     root["Hostname"] = MQTT_CLIENT_ID;
     root["Model"] = DEVICE_MODEL;
     root["IP Address"] = WiFi.localIP();
@@ -197,8 +195,9 @@ void publishToMQTT(const char* topic, const char* payload) {
     #elif defined(RF_REMOTE)
       root["Remote Type"] = "Radio Frequency (RF)";
     #endif
-    root.printTo(attributesBuf, sizeof(attributesBuf));
-    publishToMQTT(HOME_ASSISTANT_MQTT_ATTRIBUTES_TOPIC, attributesBuf);
+    char outgoingJsonBuffer[12];
+    serializeJson(root, outgoingJsonBuffer);
+    publishToMQTT(HOME_ASSISTANT_MQTT_ATTRIBUTES_TOPIC, outgoingJsonBuffer);
   }
 #endif
 
@@ -400,12 +399,14 @@ void callback(char* p_topic, byte* p_payload, unsigned int p_length) {
   // Handle the MQTT topic of the received message
   #ifdef JSON
     if (String(ARILUX_MQTT_JSON_COMMAND_TOPIC).equals(p_topic)) {
-      DynamicJsonBuffer incomingJsonPayload;
-      JsonObject& root = incomingJsonPayload.parseObject(payload);
-      if (!root.success()) {
-        DEBUG_PRINTLN("parseObject() failed");
+      DynamicJsonDocument doc(1024);
+      auto error = deserializeJson(doc, payload);
+      if (error) {
+        DEBUG_PRINT(F("parseObject() failed with code "));
+        DEBUG_PRINTLN(error.c_str());
         return;
       }
+      JsonObject root = doc.as<JsonObject>();
 
       if (root.containsKey("color")) {
         startFade = true;
@@ -531,7 +532,7 @@ void connectMQTT(void) {
         DEBUG_PRINTLN(F("INFO: The client is successfully connected to the MQTT broker"));
         publishToMQTT(ARILUX_MQTT_STATUS_TOPIC, "alive");
         #ifdef HOME_ASSISTANT_MQTT_DISCOVERY
-          JsonObject& root = HOME_ASSISTANT_MQTT_DISCOVERY_CONFIG.createObject();
+          DynamicJsonDocument root(JSON_OBJECT_SIZE(14));
           root["name"] = friendlyName;
           #ifdef HOME_ASSISTANT_MQTT_ATTRIBUTES
             root["attributes_topic"] = HOME_ASSISTANT_MQTT_ATTRIBUTES_TOPIC;
@@ -555,7 +556,7 @@ void connectMQTT(void) {
             root["payload_on"] = MQTT_STATE_ON_PAYLOAD;
             root["payload_off"] = MQTT_STATE_OFF_PAYLOAD;
           #endif
-          root.printTo(discoveryBuf, sizeof(discoveryBuf));
+          serializeJson(root, discoveryBuf);
           publishToMQTT(HOME_ASSISTANT_MQTT_DISCOVERY_TOPIC, discoveryBuf);
         #endif
         flashSuccess(true);
@@ -719,7 +720,7 @@ void handleIRRemote(void) {
         break;
       default:
         DEBUG_PRINT(F("ERROR: IR code not defined: "));
-        DEBUG_PRINTLN_WITH_FMT(results.value, HEX);
+        DEBUG_PRINTLN(uint64ToString(results.value, HEX));
         break;
     }
     irRecv.resume();
@@ -841,18 +842,18 @@ void handleRFRemote(void) {
 void handleCMD(void) {
   #ifdef JSON
   if (cmd != ARILUX_CMD_NOT_DEFINED) {
-      DynamicJsonBuffer outgoingJsonPayload;
-      JsonObject& root = outgoingJsonPayload.createObject();
+      DynamicJsonDocument root(JSON_OBJECT_SIZE(7));
       String stringState = arilux.getState() ? "ON" : "OFF";
       root["state"] = stringState;
       root["brightness"] = arilux.getBrightness();
       // root["transition"] =
       root["white_value"] = arilux.getWhite1Value();
-      JsonObject& color = root.createNestedObject("color");
+      JsonObject color = root.createNestedObject("color");
       color["r"] = arilux.getRedValue();
       color["g"] = arilux.getGreenValue();
       color["b"] = arilux.getBlueValue();
-      root.printTo(outgoingJsonBuffer);
+      char outgoingJsonBuffer[12];
+      serializeJson(root, outgoingJsonBuffer);
       publishToMQTT(ARILUX_MQTT_JSON_STATE_TOPIC, outgoingJsonBuffer);
   };
   #else

--- a/README.md
+++ b/README.md
@@ -103,7 +103,7 @@ This firmware can work with MQTT in one of two ways. To cut down on firmware siz
 1. JSON mode. Only one topic will be published and subscribed to (as well as the Last Will and Testament topic). The payload for both will be/is expected to be a JSON object with all properties listed below in it. If a property is missing, the state of that property will change.
    JSON is great because it reduces roundtrips across the network. For example, if you wanted to turn the light on, set it to full brightness, and make it red, you would have to publish to the `state` topic, the `brightness` topic and the `color` topic.
    JSON also allows effects and transitions to be specified. No properties are required. **You must have the [ArduinoJson] library installed for this to work.**
-   If you are have Home Assistant MQTT Discovery enabled, the `light.mqtt_json` platform will be loaded by Home Assistant instead of the `light.mqtt` platform.
+   If you are have Home Assistant MQTT Discovery enabled, the `light.mqtt` platform will be loaded by Home Assistant with `schema` set to `json`.
    To enable JSON mode, uncomment `#define JSON` in `config.h`.
 
    ##### JSON properties
@@ -175,6 +175,27 @@ There are a few one time steps that you need to take to get this working.
 
 From now on your device will announce itself to Home Assistant with all of the proper configuration information.
 
+#### Attributes
+
+Installing the [ArduinoJSON] library and uncommenting `HOME_ASSISTANT_MQTT_ATTRIBUTES` in your `config.h` will cause device specific attributes to be published. These can be consumed by Home Assistant
+to improve the display of the light, as well as provide some useful debugging information. The attributes payload looks like this:
+
+```json
+{
+  "BSSID": "AA:BB:CC:DD:EE:FF",
+  "Chip ID": "000AAB12",
+  "Hostname": "ARILUX000AAB12",
+  "IP Address": "192.168.1.2",
+  "LED Strip Type": "RGB",
+  "MAC Address": "GG:HH:II:JJ:KK:LL",
+  "Model": "LC01",
+  "Remote Type": "None",
+  "RSSI": -68,
+  "SSID": "my-funny-wifi-name",
+  "Telnet Logging Enabled": true
+}
+```
+
 ### Configuration for Home Assistant
 configuration.yaml
 ```yaml
@@ -188,12 +209,16 @@ mqtt:
 light:
   - platform: mqtt
     name: 'Arilux RGB Led Controller'
-    state_topic: 'rgb(w/ww)/<chipid>/state/state'
-    command_topic: 'rgb(w/ww)/<chipid>/state/set'
-    brightness_state_topic: 'rgb(w/ww)/<chipid>/brightness/state'
+    availability_topic: 'rgb(w/ww)/<chipid>/status'
     brightness_command_topic: 'rgb(w/ww)/<chipid>/brightness/set'
-    rgb_state_topic: 'rgb(w/ww)/<chipid>/color/state'
+    brightness_state_topic: 'rgb(w/ww)/<chipid>/brightness/state'
+    command_topic: 'rgb(w/ww)/<chipid>/state/set'
+    json_attributes_topic: 'rgb(w/ww)/<chipid>/attributes'
     rgb_command_topic: 'rgb(w/ww)/<chipid>/color/set'
+    rgb_state_topic: 'rgb(w/ww)/<chipid>/color/state'
+    state_topic: 'rgb(w/ww)/<chipid>/state/state'
+    white_value_command_topic: 'rgb(w/ww)/<chipid>/white/set'
+    white_value_state_topic: 'rgb(w/ww)/<chipid>/white/state'
 ```
 
 ## Todo

--- a/config.example.h
+++ b/config.example.h
@@ -75,8 +75,8 @@
 #define HOME_ASSISTANT_MQTT_DISCOVERY_PREFIX   "homeassistant"
 
 // Enable Home Assistant JSON Attributes support. Requires ArduinoJSON library to be installed.
-#define HOME_ASSISTANT_MQTT_ATTRIBUTES
-#define HOME_ASSISTANT_MQTT_ATTRIBUTES_TOPIC_TEMPLATE   "%s/attributes"
+// #define HOME_ASSISTANT_MQTT_ATTRIBUTES
+// #define HOME_ASSISTANT_MQTT_ATTRIBUTES_TOPIC_TEMPLATE   "%s/attributes"
 
 // Enable JSON. Requires ArduinoJSON library to be installed.
 // Home Assistant supports more features such as transitions, effects and flashing via JSON only.

--- a/config.example.h
+++ b/config.example.h
@@ -74,6 +74,10 @@
 #define HOME_ASSISTANT_MQTT_DISCOVERY
 #define HOME_ASSISTANT_MQTT_DISCOVERY_PREFIX   "homeassistant"
 
+// Enable Home Assistant JSON Attributes support. Requires ArduinoJSON library to be installed.
+#define HOME_ASSISTANT_MQTT_ATTRIBUTES
+#define HOME_ASSISTANT_MQTT_ATTRIBUTES_TOPIC_TEMPLATE   "%s/attributes"
+
 // Enable JSON. Requires ArduinoJSON library to be installed.
 // Home Assistant supports more features such as transitions, effects and flashing via JSON only.
 // #define JSON

